File: backend/src/application/use-cases/RegisterUserUseCase.ts
import { Email } from "../../domain/user/Email";
import { User } from "../../domain/user/User";
import { assertStrongPassword } from "../../domain/user/PasswordPolicy";
import { DuplicateEmailError } from "../../domain/user/errors/DuplicateEmailError";
import { UserRepository } from "../ports/UserRepository";
import { PasswordHasher } from "../ports/PasswordHasher";
import { IdGenerator } from "../ports/IdGenerator";

export interface RegisterUserInput {
  email: string;
  password: string;
}

export interface RegisterUserOutput {
  userId: string;
}

export class RegisterUserUseCase {
  constructor(
    private readonly userRepository: UserRepository,
    private readonly passwordHasher: PasswordHasher,
    private readonly idGenerator: IdGenerator,
  ) {}

  async execute(input: RegisterUserInput): Promise<RegisterUserOutput> {
    const email = Email.create(input.email);
    assertStrongPassword(input.password);

    const existingUser = await this.userRepository.findByEmail(email);
    if (existingUser) {
      throw new DuplicateEmailError(email.toString());
    }

    const passwordHash = await this.passwordHasher.hash(input.password);
    const user = User.register({
      id: this.idGenerator.generate(),
      email,
      passwordHash,
    });

    await this.userRepository.save(user);

    return { userId: user.id };
  }
}

File: backend-java/src/main/java/com/financepulse/engine/application/usecases/RegisterUserUseCase.java
package com.financepulse.engine.application.usecases;

import com.financepulse.engine.application.ports.IdGenerator;
import com.financepulse.engine.application.ports.PasswordHasher;
import com.financepulse.engine.application.ports.UserRepository;
import com.financepulse.engine.domain.user.Email;
import com.financepulse.engine.domain.user.PasswordPolicy;
import com.financepulse.engine.domain.user.User;
import com.financepulse.engine.domain.user.errors.DuplicateEmailException;

public class RegisterUserUseCase {

    private final UserRepository userRepository;
    private final PasswordHasher passwordHasher;
    private final IdGenerator idGenerator;

    public RegisterUserUseCase(UserRepository userRepository, PasswordHasher passwordHasher, IdGenerator idGenerator) {
        this.userRepository = userRepository;
        this.passwordHasher = passwordHasher;
        this.idGenerator = idGenerator;
    }

    public Output execute(Input input) {
        Email email = Email.create(input.email());
        PasswordPolicy.assertStrongPassword(input.password());

        if (userRepository.findByEmail(email).isPresent()) {
            throw new DuplicateEmailException(email.toString());
        }

        String passwordHash = passwordHasher.hash(input.password());
        User user = User.register(idGenerator.generate(), email, passwordHash);

        userRepository.save(user);

        return new Output(user.getId());
    }

    public record Input(String email, String password) {
    }

    public record Output(String userId) {
    }
}
